contracts/interchain-token-service/src/state.rs

```diff
@@ -107,6 +140,89 @@ pub fn load_all_its_contracts(
         .collect::<Result<HashMap<_, _>, _>>()?)
 }
 
+pub fn save_token_info(
+    storage: &mut dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+    is_origin_chain: bool,
+    track_balance: bool,
+) -> Result<(), Error> {
+    match TOKEN_INFO.may_load(storage, &(chain.clone(), token_id.clone()))? {
+        Some(_) if is_origin_chain => (),
+        Some(_) => return Err(Error::TokenNotDeployed { token_id, chain }),
+        None => {
+            let balance = TokenBalance {
+                balance: track_balance.then(Uint256::zero),
+                is_origin_chain,
+            };
+            TOKEN_INFO.save(storage, &(chain, token_id), &TokenInfo { balance })?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn update_token_info(
+    storage: &mut dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+    amount: nonempty::Uint256,
+    is_deposit: bool,
+) -> Result<(), Error> {
+    let Some(mut token_config) =
+        TOKEN_INFO.may_load(storage, &(chain.clone(), token_id.clone()))?
+    else {
+        return Err(Error::TokenNotDeployed { token_id, chain });
+    };
+
+    token_config.balance = token_config
+        .balance
+        .update(amount, is_deposit)
+        .map_err(|_| Error::TokenBalanceInvariantViolated {
+            token_id: token_id.clone(),
+            chain: chain.clone(),
+        })?;
+
+    TOKEN_INFO.save(storage, &(chain, token_id), &token_config)?;
+
+    Ok(())
+}
+
+pub fn may_load_token_info(
+    storage: &dyn Storage,
+    chain: &ChainNameRaw,
+    token_id: &TokenId,
+) -> Result<Option<TokenInfo>, Error> {
+    Ok(TOKEN_INFO.may_load(storage, &(chain.clone(), token_id.clone()))?)
+}
+
+impl TokenBalance {
+    /// Deposit an amount to the token chain balance.
+    /// If the token originated on this chain, the amount is subtracted from the balance.
+    fn update(&self, amount: nonempty::Uint256, is_deposit: bool) -> Result<Self, OverflowError> {
+        let TokenBalance {
+            balance,
+            is_origin_chain,
+        } = self;
+
+        // If the token originated on this chain, deposits/withdrawals are reversed
+        // since the balance is tracking the total supply that was bridged to other chains.
```

Comment:
this is confusing to me. What exactly are we even tracking for the origin chain? It looks like the total balance that was bridged out of the chain? Which is the reverse of what we track on the other chains (the total balance bridged to the chain?). If that is the case, why do we even need to track anything for the origin chain?